src/opuntia.erl

```diff
@@ -1,53 +1,160 @@
+-define(NON_NEG_INT(N), (is_integer(N) andalso N > 0)).
+-define(TU(T), (millisecond =:= T)).
 
--spec new(rate()) -> shaper().
+%% @doc Creates a new shaper according to the configuration.
+%%
+%% If zero is given, no shaper in created and any update action will always return zero delay;
+%% to configure a shaper it will need a
+%% ```
+%% #{bucket_size => MaximumTokens, rate => Rate, start_full => Boolean},
+%% '''
+%% where
+%% <ul>
+%%  <li>`Rate' is the number of tokens per millisecond the bucket will grow with.</li>
+%%  <li>`MaximumTokens' is the maximum number of tokens the bucket can grow.</li>
+%%  <li>`StartFull' indicates if the shaper starts with the bucket full, or empty if not.</li>
+%% </ul>
+%%
+%% So, for example, if we configure a shaper with the following:
+%% ```
+%% #{bucket_size => 60000, rate => 10, start_full => true}
+%% '''
+%% it means that the bucket will
+%% allow `10' tokens per `millisecond', up to 60000 tokens, regardless of how long it is left
+%% unused to charge: it will never charge further than 60000 tokens.
+-spec new(config()) -> shaper().
 new(0) ->
     none;
-new(MaxRatePerMs) ->
-    #token_bucket{rate = MaxRatePerMs,
-                  available_tokens = MaxRatePerMs,
-                  last_update = erlang:monotonic_time(millisecond)}.
+new(Shape) ->
+    create(Shape, erlang:monotonic_time()).
 
 %% @doc Update shaper and return possible waiting time.
 %%
 %% This function takes the current shaper state, and the number of tokens that have been consumed,
 %% and returns a tuple containing the new shaper state, and a possibly non-zero number of
 %% unit times to wait if more tokens that the shaper allows were consumed.
--spec update(shaper(), tokens()) -> {shaper(), rate()}.
-update(none, _Size) ->
+-spec update(shaper(), tokens()) -> {shaper(), delay()}.
+update(none, _TokensNowUsed) ->
     {none, 0};
-update(#token_bucket{rate = MaxRatePerMs,
-                     available_tokens = LastAvailableTokens,
-                     last_update = LastUpdate} = Shaper, TokensNowUsed) ->
-    Now = erlang:monotonic_time(millisecond),
-    % How much we might have recovered since last time
-    TimeSinceLastUpdate = Now - LastUpdate,
-    PossibleTokenGrowth = round(MaxRatePerMs * TimeSinceLastUpdate),
-    % Available plus recovered cannot grow higher than the actual rate limit
-    ExactlyAvailableNow = min(MaxRatePerMs, LastAvailableTokens + PossibleTokenGrowth),
-    % Now check how many tokens are available by substracting how many where used,
-    % and how many where overused
+update(Shaper, TokensNowUsed) ->
+    calculate(Shaper, TokensNowUsed, erlang:monotonic_time()).
+
+%% Helpers
+-spec create(config(), integer()) -> shaper().
+create(0, _) ->
+    none;
+create(#{bucket_size := MaximumTokens,
+         rate := Rate,
+         start_full := StartFull},
+       NativeNow)
+  when ?NON_NEG_INT(MaximumTokens),
+       ?NON_NEG_INT(Rate),
+       MaximumTokens >= Rate,
+       ?TU(millisecond),
```

Comment:
This check doesn't make much sense.

Reply:
Artifact of removing all the other code, removing this too 👌🏽